js/src/ui/Form/AddressSelect/addressSelect.js

```diff
@@ -15,262 +15,799 @@
+      this.setState({ focusedCat: nextCategory });
+      nextFocusIndex = Math.min(prevFocusIndex, values[nextCategory].values.length - 1);
+    }
+
+    const nextFocus = values[nextCategory].values[nextFocusIndex].index;
+    return this.focusItem(nextFocus);
+  }
+
+  handleClick = (address) => {
+    // Don't do anything if it's only text-selection
+    if (window.getSelection && window.getSelection().type === 'Range') {
+      return;
     }
 
     this.props.onChange(null, address);
+    this.handleClose();
+  }
+
+  handleMainBlur = () => {
+    if (window.document.hasFocus() && !this.state.expanded) {
+      this.closing = false;
+      this.setState({ focused: false });
+    }
+  }
+
+  handleMainFocus = () => {
+    if (this.state.focused) {
+      return;
+    }
+
+    this.setState({ focused: true }, () => {
+      if (this.closing) {
+        this.closing = false;
+        return;
+      }
+
+      this.handleFocus();
+    });
+  }
+
+  setPosition = (nextState = {}) => {
+    const { top = 0, left = 0 } = this.handleDOMAction(this.refs.inputAddress, 'getBoundingClientRect') || {};
+    this.setState({ top, left, ...nextState });
+  }
+
+  handleFocus = () => {
+    this.setState({ expanded: true, focusedItem: null, focusedCat: null, top: 0, left: 0 }, () => {
+      window.setTimeout(() => {
+        this.handleDOMAction(this.inputRef, 'focus');
+      });
+    });
+  }
+
+  handleClose = () => {
+    if (!this.refs.inputAddress) {
+      return null;
+    }
+
+    this.setPosition({ expanded: false });
+    this.closing = true;
+
+    return this.handleDOMAction('inputAddress', 'focus');
+  }
+
+  /**
+   * Filter the given values based on the given
+   * filter
+   */
+  filterValues = (values = [], _filter = '') => {
+    const filter = _filter.toLowerCase();
+
+    return values
+      // Remove empty accounts
+      .filter((a) => a)
+      .filter((account) => {
+        const address = account.address.toLowerCase();
+        const inAddress = address.includes(filter);
+
+        if (!account.name || inAddress) {
+          return inAddress;
+        }
+
+        const name = account.name.toLowerCase();
+        const inName = name.includes(filter);
+        const { meta = {} } = account;
+
+        if (!meta.tags || inName) {
+          return inName;
+        }
+
+        const tags = (meta.tags || []).join('');
+        return tags.includes(filter);
+      })
+      .sort((accA, accB) => {
+        const nameA = accA.name || accA.address;
+        const nameB = accB.name || accB.address;
+
+        return nameA.localeCompare(nameB);
+      });
+  }
+
+  handleInputBlur = () => {
+    this.setState({ inputFocused: false });
+  }
+
+  handleInputFocus = () => {
+    this.setState({ focusedItem: null, inputFocused: true });
+  }
+
+  handleChange = (event = { target: {} }) => {
+    const { value = '' } = event.target;
+    let index = 0;
+
+    window.setTimeout(() => { this.handleTagsOpacity(); });
```

Comment:
As per all the tag comments, edge case, not fitting in, not componentizable as it stands and due for a repeat everywhere.